File: mycode/core/agent.py
"""Agent loop with provider-specific adapters and one internal message model."""

from __future__ import annotations

import asyncio
import logging
from collections.abc import AsyncIterator, Awaitable, Callable
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

from mycode.core.config import Settings, get_settings
from mycode.core.messages import (
    ConversationMessage,
    build_message,
    flatten_message_text,
    tool_result_block,
    user_text_message,
)
from mycode.core.providers import get_provider_adapter
from mycode.core.providers.base import ProviderAdapter, ProviderRequest, ProviderStreamEvent
from mycode.core.session import (
    COMPACT_SUMMARY_PROMPT,
    DEFAULT_COMPACT_THRESHOLD,
    apply_compact,
    build_compact_event,
    should_compact,
)
from mycode.core.system_prompt import build_system_prompt
from mycode.core.tools import ToolExecutionResult, ToolExecutor

logger = logging.getLogger(__name__)

PersistCallback = Callable[[dict[str, Any]], Awaitable[None]]


@dataclass
class Event:
    """Streaming event emitted by the agent."""

    type: str
    data: dict[str, Any] = field(default_factory=dict)


class Agent:
    """Minimal coding agent with one internal loop and provider adapters."""

    def __init__(
        self,
        *,
        model: str,
        cwd: str,
        session_dir: Path,
        session_id: str | None = None,
        provider: str | None = None,
        api_key: str | None = None,
        api_base: str | None = None,
        messages: list[dict[str, Any]] | None = None,
        max_turns: int | None = None,
        max_tokens: int = 16_384,
        context_window: int | None = 128_000,
        compact_threshold: float | None = None,
        reasoning_effort: str | None = None,
        supports_image_input: bool | None = None,
        settings: Settings | None = None,
        system: str | None = None,
        tool_executor: ToolExecutor | None = None,
    ):
        self.model = model
        self.provider = provider or "anthropic"
        self.cwd = str(Path(cwd).resolve(strict=False))
        self.session_dir = session_dir
        self.session_id = (session_id or session_dir.name).strip() or None
        self.api_key = api_key
        self.api_base = api_base
        self.max_turns = max_turns
        self.max_tokens = max_tokens
        self.context_window = context_window
        self.compact_threshold = compact_threshold if compact_threshold is not None else DEFAULT_COMPACT_THRESHOLD
        self.reasoning_effort = reasoning_effort
        self.supports_image_input: bool = bool(supports_image_input)
        self.settings = settings or get_settings(self.cwd)
        self.system = system or build_system_prompt(self.cwd, self.settings)
        self._cancel_event = asyncio.Event()
        self._provider_event_task: asyncio.Future[ProviderStreamEvent] | None = None

        self.messages: list[ConversationMessage] = list(messages or [])
        self.tools = tool_executor or ToolExecutor(cwd=self.cwd, session_dir=self.session_dir)
        self.tools.supports_image_input = self.supports_image_input

    def cancel(self) -> None:
        self._cancel_event.set()
        self.tools.cancel_active()
        if self._provider_event_task and not self._provider_event_task.done():
            self._provider_event_task.cancel()

    def clear(self) -> None:
        self.messages = []

    @staticmethod
    def _tool_done_event(tool_id: str, result: ToolExecutionResult) -> Event:
        """Build the standard tool_done event payload."""

        data = {
            "tool_use_id": tool_id,
            "model_text": result.model_text,
            "display_text": result.display_text,
            "is_error": result.is_error,
        }
        if result.content:
            data["content"] = result.content
        return Event(
            "tool_done",
            data,
        )

    async def _run_streaming_tool(self, *, tool_id: str, name: str, args: dict[str, Any]) -> AsyncIterator[Event]:
        """Run one streaming tool and forward live output until it finishes."""

        loop = asyncio.get_running_loop()
        output_queue: asyncio.Queue[str | None] = asyncio.Queue()

        def on_output(line: str) -> None:
            loop.call_soon_threadsafe(output_queue.put_nowait, line)

        async def run_in_thread() -> ToolExecutionResult:
            try:
                return await asyncio.to_thread(
                    self.tools.run_streaming,
                    name,
                    tool_call_id=tool_id,
                    args=args,
                    on_output=on_output,
                )
            finally:
                loop.call_soon_threadsafe(output_queue.put_nowait, None)

        task = asyncio.create_task(run_in_thread())
        was_cancelled = False

        # Streaming tools produce intermediate output, but they still end as one
        # normal tool_done event so the outer loop can handle all tools the same way.
        while True:
            if self._cancel_event.is_set() and not was_cancelled:
                was_cancelled = True
                self.tools.cancel_active()

            try:
                output = await asyncio.wait_for(output_queue.get(), timeout=0.1)
            except TimeoutError:
                if task.done():
                    break
                continue

            if output is None:
                break
            if not was_cancelled:
                yield Event("tool_output", {"tool_use_id": tool_id, "output": output})

        if was_cancelled:
            try:
                await task
            except Exception:
                pass
            result = ToolExecutionResult(
                model_text="error: cancelled",
                display_text="Cancelled",
                is_error=True,
            )
        else:
            try:
                result = await task
            except Exception as exc:  # pragma: no cover - defensive
                result = ToolExecutionResult(
                    model_text=f"error: {exc}",
                    display_text=str(exc),
                    is_error=True,
                )

        yield self._tool_done_event(tool_id, result)

    async def _run_tool_call(self, tool_use: dict[str, Any]) -> AsyncIterator[Event]:
        """Run one tool call and emit the standard tool events."""

        tool_id = str(tool_use.get("id") or "")
        name = str(tool_use.get("name") or "")
        raw_args = tool_use.get("input")
        args = raw_args if isinstance(raw_args, dict) else {}

        yield Event("tool_start", {"tool_call": {"id": tool_id, "name": name, "input": args}})

        if self._cancel_event.is_set():
            yield self._tool_done_event(
                tool_id,
                ToolExecutionResult(
                    model_text="error: cancelled",
                    display_text="Cancelled",
                    is_error=True,
                ),
            )
            return

        tool = self.tools.get_tool(name)
        if tool is None:
            yield self._tool_done_event(
                tool_id,
                ToolExecutionResult(
                    model_text=f"error: unknown tool: {name}",
                    display_text=f"Unknown tool: {name}",
                    is_error=True,
                ),
            )
            return

        if tool.streams_output:
            async for event in self._run_streaming_tool(tool_id=tool_id, name=name, args=args):
                yield event
            return

        try:
            result = self.tools.run(name, args=args)
        except Exception as exc:  # pragma: no cover - defensive
            result = ToolExecutionResult(
                model_text=f"error: {exc}",
                display_text=str(exc),
                is_error=True,
            )

        yield self._tool_done_event(tool_id, result)

    async def _stream_provider_turn(
        self,
        adapter: ProviderAdapter,
        request: ProviderRequest,
    ) -> AsyncIterator[ProviderStreamEvent]:
        """Iterate one provider turn with best-effort cancellation support."""

        provider_stream: AsyncIterator[ProviderStreamEvent] = adapter.stream_turn(request)
        try:
            while True:
                if self._cancel_event.is_set():
                    raise asyncio.CancelledError

                self._provider_event_task = asyncio.create_task(anext(provider_stream))
                try:
                    yield await self._provider_event_task
                except StopAsyncIteration:
                    return
                finally:
                    self._provider_event_task = None
        finally:
            close = getattr(provider_stream, "aclose", None)
            if callable(close):
                try:
                    await close()
                except Exception:
                    pass

    async def achat(
        self,
        user_input: str | ConversationMessage,
        *,
        on_persist: PersistCallback | None = None,
    ) -> AsyncIterator[Event]:
        """Run the full agent loop for one user message.

        Each turn asks the provider for one assistant message. If the assistant
        requests tools, the agent runs them locally, appends one user-side
        tool_result message, and continues until the assistant stops using tools.
        """

        self._cancel_event.clear()
        supports_image_input = self.supports_image_input
        self.tools.supports_image_input = supports_image_input

        if isinstance(user_input, str):
            user_message = user_text_message(user_input)
        else:
            user_message = {
                "role": str(user_input.get("role") or "user"),
                "content": [dict(b) for b in user_input.get("content") or [] if isinstance(b, dict)],
            }
            if isinstance(user_input.get("meta"), dict):
                user_message["meta"] = dict(user_input["meta"])

        if user_message.get("role") != "user":
            yield Event("error", {"message": "user input must be a user message"})
            return

        if not supports_image_input and any(
            isinstance(block, dict) and block.get("type") == "image" for block in user_message.get("content") or []
        ):
            yield Event("error", {"message": "current model does not support image input"})
            return

        self.messages.append(user_message)
        if on_persist:
            await on_persist(user_message)

        adapter = get_provider_adapter(self.provider)

        turn_number = 0
        while self.max_turns is None or turn_number < self.max_turns:
            turn_number += 1
            if self._cancel_event.is_set():
                yield Event("error", {"message": "cancelled"})
                return

            assistant_message: ConversationMessage | None = None
            request = ProviderRequest(
                provider=self.provider,
                model=self.model,
                session_id=self.session_id,
                messages=self.messages,
                system=self.system,
                tools=self.tools.definitions,
                max_tokens=self.max_tokens,
                api_key=self.api_key,
                api_base=self.api_base,
                reasoning_effort=self.reasoning_effort,
                supports_image_input=supports_image_input,
            )

            try:
                # Phase 1: ask the provider for exactly one assistant turn.
                async for provider_event in self._stream_provider_turn(adapter, request):
                    if self._cancel_event.is_set():
                        yield Event("error", {"message": "cancelled"})
                        return

                    if provider_event.type == "thinking_delta":
                        delta_text = str(provider_event.data.get("text") or "")
                        if delta_text:
                            yield Event("reasoning", {"delta": delta_text})
                        continue

                    if provider_event.type == "text_delta":
                        delta_text = str(provider_event.data.get("text") or "")
                        if delta_text:
                            yield Event("text", {"delta": delta_text})
                        continue

                    if provider_event.type == "provider_error":
                        raise ValueError(str(provider_event.data.get("message") or "provider error"))

                    if provider_event.type != "message_done":
                        continue

                    message = provider_event.data.get("message")
                    if isinstance(message, dict):
                        assistant_message = message

            except asyncio.CancelledError:
                yield Event("error", {"message": "cancelled"})
                return
            except Exception as exc:
                logger.exception("Provider request failed")
                yield Event("error", {"message": str(exc)})
                return

            if not assistant_message:
                yield Event("error", {"message": "provider produced no assistant message"})
                return

            self.messages.append(assistant_message)
            if on_persist:
                await on_persist(assistant_message)

            # Phase 2: if the assistant requested tools, execute them locally and
            # append one user-side tool_result message before continuing.
            tool_calls = [
                block
                for block in assistant_message.get("content") or []
                if isinstance(block, dict) and block.get("type") == "tool_use"
            ]
            if not tool_calls:
                break

            tool_results: list[dict[str, Any]] = []
            for tool_call in tool_calls:
                async for event in self._run_tool_call(tool_call):
                    yield event

                    if event.type != "tool_done":
                        continue

                    d = event.data
                    model_text = str(d.get("model_text") or "")
                    content = d.get("content")
                    tool_results.append(
                        tool_result_block(
                            tool_use_id=str(d.get("tool_use_id") or ""),
                            model_text=model_text,
                            display_text=str(d.get("display_text") or ""),
                            is_error=bool(d.get("is_error")),
                            content=content if isinstance(content, list) else None,
                        )
                    )

                    if model_text == "error: cancelled" and self._cancel_event.is_set():
                        tool_result_message = build_message("user", tool_results)
                        self.messages.append(tool_result_message)
                        if on_persist:
                            await on_persist(tool_result_message)
                        return

            tool_result_message = build_message("user", tool_results)
            self.messages.append(tool_result_message)
            if on_persist:
                await on_persist(tool_result_message)

        else:
            # while loop exhausted max_turns without breaking
            yield Event("error", {"message": "max_turns reached"})
            return

        # Turn completed normally (assistant stopped calling tools).
        # Check whether context compaction is needed.
        if not self._cancel_event.is_set():
            async for event in self._compact_if_needed(adapter, on_persist):
                yield event

    # -----------------------------------------------------------------
    # Context compaction
    # -----------------------------------------------------------------

    async def _compact_if_needed(
        self,
        adapter: ProviderAdapter,
        on_persist: PersistCallback | None,
    ) -> AsyncIterator[Event]:
        """Check token usage and run compaction if above threshold."""

        usage: dict[str, Any] | None = None
        for message in reversed(self.messages):
            if message.get("role") == "assistant":
                usage = (message.get("meta") or {}).get("usage")
                break

        if not should_compact(usage, self.context_window, self.compact_threshold):
            return

        try:
            async for event in self._compact(adapter, on_persist):
                yield event
        except (Exception, asyncio.CancelledError):
            logger.warning("Context compaction failed, continuing without compaction", exc_info=True)

    async def _compact(
        self,
        adapter: ProviderAdapter,
        on_persist: PersistCallback | None,
    ) -> AsyncIterator[Event]:
        """Generate a conversation summary and replace in-memory messages."""
        compacted_count = len(self.messages)

        # Ask the same provider for a summary — no tools, just text generation.
        compact_messages = list(self.messages) + [user_text_message(COMPACT_SUMMARY_PROMPT)]
        request = ProviderRequest(
            provider=self.provider,
            model=self.model,
            session_id=self.session_id,
            messages=compact_messages,
            system=self.system,
            tools=[],
            max_tokens=min(self.max_tokens, 8192),
            api_key=self.api_key,
            api_base=self.api_base,
            supports_image_input=self.supports_image_input,
        )

        summary_message: ConversationMessage | None = None
        async for provider_event in self._stream_provider_turn(adapter, request):
            if provider_event.type == "message_done":
                msg = provider_event.data.get("message")
                if isinstance(msg, dict):
                    summary_message = msg

        if not summary_message:
            logger.warning("Compaction produced no response")
            return

        summary_text = flatten_message_text(summary_message, include_thinking=False)
        if not summary_text:
            logger.warning("Compaction produced empty summary")
            return

        summary_usage = (summary_message.get("meta") or {}).get("usage")
        compact_event = build_compact_event(
            summary_text,
            provider=self.provider,
            model=self.model,
            compacted_count=compacted_count,
            usage=summary_usage,
        )

        # Persist the compact event (append-only — original messages stay in JSONL).
        if on_persist:
            await on_persist(compact_event)

        # Rebuild in-memory messages from the compact event.
        self.messages.append(compact_event)
        self.messages = apply_compact(self.messages)

        yield Event(
            "compact",
            {
                "message": f"Context compacted ({compacted_count} messages → summary)",
                "compacted_count": compacted_count,
            },
        )

File: mycode-go/internal/agent/agent.go
package agent

import (
	"context"
	"errors"
	"fmt"
	"maps"
	"path/filepath"

	"github.com/legibet/mycode-go/internal/config"
	"github.com/legibet/mycode-go/internal/message"
	"github.com/legibet/mycode-go/internal/prompt"
	"github.com/legibet/mycode-go/internal/provider"
	"github.com/legibet/mycode-go/internal/session"
	"github.com/legibet/mycode-go/internal/tools"
)

// Event is the normalized streaming event sent to the API and CLI.
type Event struct {
	Type string
	Data map[string]any
}

// PersistFunc stores one canonical message.
type PersistFunc func(message.Message) error

// Agent is the single orchestration loop.
type Agent struct {
	Model              string
	Provider           string
	CWD                string
	SessionDir         string
	SessionID          string
	APIKey             string
	APIBase            string
	MaxTurns           int
	MaxTokens          int
	ContextWindow      int
	CompactThreshold   float64
	ReasoningEffort    string
	SupportsImageInput bool
	SupportsPDFInput   bool
	System             string
	Messages           []message.Message
	Tools              *tools.Executor
	Adapter            provider.Adapter
}

// New creates an agent.
func New(
	model, providerType, cwd, sessionDir, sessionID, apiKey, apiBase, system string,
	messages []message.Message,
	maxTurns, maxTokens, contextWindow int,
	compactThreshold float64,
	reasoningEffort string,
	supportsImageInput, supportsPDFInput bool,
	adapter provider.Adapter,
	toolExecutor *tools.Executor,
) (*Agent, error) {
	resolvedSessionDir := sessionDir
	if resolvedSessionDir == "" {
		resolvedSessionDir = cwd
	}
	resolvedSessionID := sessionID
	if resolvedSessionID == "" {
		resolvedSessionID = filepath.Base(resolvedSessionDir)
	}
	if toolExecutor == nil {
		toolExecutor = tools.NewExecutor(cwd, resolvedSessionDir, supportsImageInput)
	}
	if adapter == nil {
		var ok bool
		adapter, ok = provider.LookupAdapter(providerType)
		if !ok {
			return nil, errors.New("unsupported provider adapter: " + providerType)
		}
	}
	if system == "" {
		system = prompt.Build(cwd, config.ResolveHome())
	}
	cloned := make([]message.Message, len(messages))
	for i, msg := range messages {
		cloned[i] = message.Clone(msg)
	}
	return &Agent{
		Model:              model,
		Provider:           providerType,
		CWD:                cwd,
		SessionDir:         resolvedSessionDir,
		SessionID:          resolvedSessionID,
		APIKey:             apiKey,
		APIBase:            apiBase,
		MaxTurns:           maxTurns,
		MaxTokens:          maxTokens,
		ContextWindow:      contextWindow,
		CompactThreshold:   compactThreshold,
		ReasoningEffort:    reasoningEffort,
		SupportsImageInput: supportsImageInput,
		SupportsPDFInput:   supportsPDFInput,
		System:             system,
		Messages:           cloned,
		Tools:              toolExecutor,
		Adapter:            adapter,
	}, nil
}

// Cancel stops active tools. Provider cancellation is driven by ctx.
func (a *Agent) Cancel() {
	a.Tools.CancelActive()
}

// Chat runs one user turn.
func (a *Agent) Chat(ctx context.Context, userInput message.Message, onPersist PersistFunc) <-chan Event {
	out := make(chan Event, 32)
	go func() {
		defer close(out)
		if userInput.Role != "user" {
			out <- Event{Type: "error", Data: map[string]any{"message": "user input must be a user message"}}
			return
		}
		if err := a.validateUserInput(userInput); err != nil {
			out <- Event{Type: "error", Data: map[string]any{"message": err.Error()}}
			return
		}

		a.Messages = append(a.Messages, message.Clone(userInput))
		if onPersist != nil {
			if err := onPersist(userInput); err != nil {
				out <- Event{Type: "error", Data: map[string]any{"message": err.Error()}}
				return
			}
		}

		turn := 0
		completed := false
		for a.MaxTurns <= 0 || turn < a.MaxTurns {
			turn++
			req := provider.Request{
				Provider:           a.Provider,
				Model:              a.Model,
				SessionID:          a.SessionID,
				Messages:           a.Messages,
				System:             a.System,
				Tools:              toolSpecs(a.Tools.Definitions()),
				MaxTokens:          a.MaxTokens,
				APIKey:             a.APIKey,
				APIBase:            a.APIBase,
				ReasoningEffort:    a.ReasoningEffort,
				SupportsImageInput: a.SupportsImageInput,
				SupportsPDFInput:   a.SupportsPDFInput,
			}

			var assistant *message.Message
			for event := range a.Adapter.StreamTurn(ctx, req) {
				switch event.Type {
				case "thinking_delta":
					if event.Text != "" {
						out <- Event{Type: "reasoning", Data: map[string]any{"delta": event.Text}}
					}
				case "text_delta":
					if event.Text != "" {
						out <- Event{Type: "text", Data: map[string]any{"delta": event.Text}}
					}
				case "message_done":
					assistant = event.Msg
				case "provider_error":
					if event.Err != nil {
						out <- Event{Type: "error", Data: map[string]any{"message": event.Err.Error()}}
					} else {
						out <- Event{Type: "error", Data: map[string]any{"message": "provider error"}}
					}
					return
				}
			}
			if ctx.Err() != nil {
				out <- Event{Type: "error", Data: map[string]any{"message": "cancelled"}}
				return
			}
			if assistant == nil {
				out <- Event{Type: "error", Data: map[string]any{"message": "provider produced no assistant message"}}
				return
			}

			a.Messages = append(a.Messages, message.Clone(*assistant))
			if onPersist != nil {
				if err := onPersist(*assistant); err != nil {
					out <- Event{Type: "error", Data: map[string]any{"message": err.Error()}}
					return
				}
			}

			toolCalls := make([]message.Block, 0)
			for _, block := range assistant.Content {
				if block.Type == "tool_use" {
					toolCalls = append(toolCalls, block)
				}
			}
			if len(toolCalls) == 0 {
				completed = true
				break
			}

			toolResults := make([]message.Block, 0, len(toolCalls))
			for _, toolCall := range toolCalls {
				select {
				case <-ctx.Done():
					out <- Event{Type: "error", Data: map[string]any{"message": "cancelled"}}
					return
				default:
				}

				out <- Event{Type: "tool_start", Data: map[string]any{
					"tool_call": map[string]any{
						"id":    toolCall.ID,
						"name":  toolCall.Name,
						"input": cloneInput(toolCall.Input),
					},
				}}

				result := a.runTool(toolCall, out)
				toolResults = append(toolResults, message.ToolResultBlock(
					toolCall.ID,
					result.ModelText,
					result.DisplayText,
					result.IsError,
					result.Content,
					nil,
				))

				data := map[string]any{
					"tool_use_id":  toolCall.ID,
					"model_text":   result.ModelText,
					"display_text": result.DisplayText,
					"is_error":     result.IsError,
				}
				if len(result.Content) > 0 {
					data["content"] = result.Content
				}
				out <- Event{Type: "tool_done", Data: data}

				if result.ModelText == "error: cancelled" && ctx.Err() != nil {
					toolMessage := message.BuildMessage("user", toolResults, nil)
					a.Messages = append(a.Messages, toolMessage)
					if onPersist != nil {
						_ = onPersist(toolMessage)
					}
					return
				}
			}

			toolMessage := message.BuildMessage("user", toolResults, nil)
			a.Messages = append(a.Messages, toolMessage)
			if onPersist != nil {
				if err := onPersist(toolMessage); err != nil {
					out <- Event{Type: "error", Data: map[string]any{"message": err.Error()}}
					return
				}
			}
		}
		if !completed && a.MaxTurns > 0 {
			out <- Event{Type: "error", Data: map[string]any{"message": "max_turns reached"}}
			return
		}

		for event := range a.compactIfNeeded(ctx, onPersist) {
			out <- event
		}
	}()
	return out
}

func (a *Agent) validateUserInput(userInput message.Message) error {
	for _, block := range userInput.Content {
		if block.Type == "image" && !a.SupportsImageInput {
			return fmt.Errorf("current model does not support image input")
		}
		if block.Type == "document" && !a.SupportsPDFInput {
			return fmt.Errorf("current model does not support PDF input")
		}
	}
	return nil
}

func (a *Agent) runTool(toolCall message.Block, out chan<- Event) tools.Result {
	switch toolCall.Name {
	case "read":
		return a.Tools.Read(asString(toolCall.Input["path"]), asInt(toolCall.Input["offset"]), asInt(toolCall.Input["limit"]))
	case "write":
		return a.Tools.Write(asString(toolCall.Input["path"]), asString(toolCall.Input["content"]))
	case "edit":
		return a.Tools.Edit(asString(toolCall.Input["path"]), asEdits(toolCall.Input["edits"]))
	case "bash":
		return a.Tools.Bash(toolCall.ID, asString(toolCall.Input["command"]), asInt(toolCall.Input["timeout"]), func(text string) {
			out <- Event{Type: "tool_output", Data: map[string]any{
				"tool_use_id": toolCall.ID,
				"output":      text,
			}}
		})
	default:
		return tools.Result{ModelText: "error: unknown tool: " + toolCall.Name, DisplayText: "Unknown tool: " + toolCall.Name, IsError: true}
	}
}

func (a *Agent) compactIfNeeded(ctx context.Context, onPersist PersistFunc) <-chan Event {
	out := make(chan Event, 1)
	go func() {
		defer close(out)
		if len(a.Messages) == 0 {
			return
		}

		var usage map[string]any
		for i := len(a.Messages) - 1; i >= 0; i-- {
			msg := a.Messages[i]
			if msg.Role != "assistant" {
				continue
			}
			if raw, ok := msg.Meta["usage"].(map[string]any); ok {
				usage = raw
			}
			break
		}
		if !session.ShouldCompact(usage, a.ContextWindow, a.CompactThreshold) {
			return
		}

		beforeCount := len(a.Messages)
		compactMessages := append(append([]message.Message(nil), a.Messages...), message.UserTextMessage(session.CompactSummaryPrompt, nil))
		req := provider.Request{
			Provider:           a.Provider,
			Model:              a.Model,
			SessionID:          a.SessionID,
			Messages:           compactMessages,
			System:             a.System,
			MaxTokens:          min(a.MaxTokens, 8192),
			APIKey:             a.APIKey,
			APIBase:            a.APIBase,
			SupportsImageInput: a.SupportsImageInput,
			SupportsPDFInput:   a.SupportsPDFInput,
		}

		var summary *message.Message
		for event := range a.Adapter.StreamTurn(ctx, req) {
			if event.Type == "message_done" {
				summary = event.Msg
			}
			if event.Type == "provider_error" {
				return
			}
		}
		if ctx.Err() != nil || summary == nil {
			return
		}
		summaryText := message.FlattenText(*summary, false)
		if summaryText == "" {
			return
		}
		compactEvent := session.BuildCompactEvent(summaryText, a.Provider, a.Model, beforeCount, summary.Meta["usage"])
		if onPersist != nil {
			if err := onPersist(compactEvent); err != nil {
				return
			}
		}
		a.Messages = append(a.Messages, compactEvent)
		a.Messages = session.ApplyCompact(a.Messages)
		out <- Event{Type: "compact", Data: map[string]any{
			"message":         fmt.Sprintf("Context compacted (%d messages -> summary)", beforeCount),
			"compacted_count": beforeCount,
		}}
	}()
	return out
}

// toolSpecs converts internal tool definitions to provider-facing maps.
func toolSpecs(specs []tools.ToolSpec) []map[string]any {
	out := make([]map[string]any, 0, len(specs))
	for _, spec := range specs {
		out = append(out, map[string]any{
			"name":         spec.Name,
			"description":  spec.Description,
			"input_schema": spec.InputSchema,
		})
	}
	return out
}

func cloneInput(input map[string]any) map[string]any {
	if input == nil {
		return map[string]any{}
	}
	return maps.Clone(input)
}

func asString(value any) string {
	text, _ := value.(string)
	return text
}

func asInt(value any) int {
	switch v := value.(type) {
	case int:
		return v
	case float64:
		return int(v)
	default:
		return 0
	}
}

func asEdits(value any) []map[string]string {
	items, _ := value.([]any)
	out := make([]map[string]string, 0, len(items))
	for _, item := range items {
		entry, _ := item.(map[string]any)
		if entry == nil {
			continue
		}
		out = append(out, map[string]string{
			"oldText": asString(entry["oldText"]),
			"newText": asString(entry["newText"]),
		})
	}
	return out
}
